test/test_optim.py

```diff
@@ -1647,6 +1649,7 @@ def test_functional_fused_optimizer_with_foundinf(self):
                     for _ in range(num_tensors)
                 ],
             )
+            self.assertEqual(params, prev_params)
```

Comment:
If I understand properly, this test case tests that the the params are NOT mutated for functional optimizers, yes? If so, should we add another test that verifies the bug would not regress?

Reply:
The cause is that `found_inf` is not passed to the kernels when `grad_scale` is not set to the fused optimizers thus the parameters got updated using invalid gradients, leading to nan loss.
So one appropriate test is to make sure that the parameters are not updated after `optimizer.step` call if `found_inf` is `1` and there are some invalid gradients.

Reply:
Ah lol my bad, I misinterpreted "functional" there. This test + fix make sense together.

Reply:
Are there any docs on when grad_scale is set vs not set? In the issue, it looks like the bug repros had turned GradScaler to on since they're using float16, and this seems to catch a case when grad_scale is off. I'm going to approve since this fix seems like something we definitely want! I'd also like to understand AMP more and verify the bug would be completely fixed.

Reply:
when GradScaler is employed, 
- found_inf: needs to be set always
- grad_scale: needs to be set unless we manually call `GradScaler.unscale_`. A manual call of `GradScaler.unscale_` is sort of necessary to modify gradients, e.g. https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_norm_.html